File: 04_nvidia/daxpy_acc.cpp
#include <cstdlib>

int main() {
    double* x;
    double* y;

    const int N = 1000000;
    const double a = 3.0;

    x = (double*) malloc(N * sizeof(double));
    y = (double*) malloc(N * sizeof(double));

    for (int i = 0; i < N; ++i) {
        x[i] = 1.0 * i;
        y[i] = 2.0 * i;
    }

#pragma acc parallel loop
    for (int i = 0; i < N; ++i) {
        y[i] = a * x[i] + y[i];
    }

    free(x);
    free(y);
}

File: 04_nvidia/daxpy_acc.f90
program main

  implicit none

  integer, parameter :: N = 1000000
  double precision, parameter :: a = 3.0
  double precision, allocatable :: x(:), y(:)
  integer :: i

  allocate(x(N))
  allocate(y(N))

  do i = 1, N
     x(i) = 1.0 * i
     y(i) = 2.0 * i
  end do

  !$acc parallel loop
  do i = 1, N
     y(i) = a * x(i) + y(i)
  end do

end program main
